test/cert_comp_test.c

```diff
@@ -0,0 +1,280 @@
+
+static char *certsdir = NULL;
+static char *cert = NULL;
+static char *privkey = NULL;
+
+static int client_cert_cb(SSL *ssl, X509 **x509, EVP_PKEY **pkey)
+{
+    X509 *xcert;
+    EVP_PKEY *privpkey;
+    BIO *in = NULL;
+    BIO *priv_in = NULL;
+
+    /* Check that SSL_get0_peer_certificate() returns something sensible */
+    if (!TEST_ptr(SSL_get0_peer_certificate(ssl)))
+        return 0;
+
+    in = BIO_new_file(cert, "r");
+    if (!TEST_ptr(in))
+        return 0;
+
+    if (!TEST_ptr(xcert = X509_new_ex(NULL, NULL))
+            || !TEST_ptr(PEM_read_bio_X509(in, &xcert, NULL, NULL))
+            || !TEST_ptr(priv_in = BIO_new_file(privkey, "r"))
+            || !TEST_ptr(privpkey = PEM_read_bio_PrivateKey_ex(priv_in, NULL,
+                                                               NULL, NULL,
+                                                               NULL, NULL)))
+        goto err;
+
+    *x509 = xcert;
+    *pkey = privpkey;
+
+    BIO_free(in);
+    BIO_free(priv_in);
+    return 1;
+err:
+    X509_free(xcert);
+    BIO_free(in);
+    BIO_free(priv_in);
+    return 0;
+}
+
+static int verify_cb(int preverify_ok, X509_STORE_CTX *x509_ctx)
+{
+    return 1;
+}
+
+/*
+ * Test 0 = app pre-compresses certificate in SSL
+ * Test 1 = app pre-compresses certificate in SSL_CTX
+ * Test 2 = app pre-compresses certificate in SSL_CTX, client authentication
+ * Test 3 = app pre-compresses certificate in SSL_CTX, but it's unused due to prefs
+ */
+/* Compression helper */
+static int ssl_comp_cert(SSL *ssl, int alg)
+{
+    unsigned char *comp_data = NULL;
+    size_t comp_len = 0;
+    size_t orig_len = 0;
+    int retval = 0;
+
+    if (!TEST_size_t_gt(comp_len = SSL_get1_compressed_cert(ssl, alg, &comp_data, &orig_len), 0))
+        goto err;
+
+    if (!TEST_true(SSL_set1_compressed_cert(ssl, alg, comp_data, comp_len, orig_len)))
+        goto err;
+    retval = alg;
+
+ err:
+    OPENSSL_free(comp_data);
+    return retval;
+}
+
+static void cert_comp_info_cb(const SSL *s, int where, int ret)
+{
+    int *seen = (int*)SSL_get_app_data(s);
+
+    if (SSL_is_server(s)) {
+        /* TLS_ST_SR_COMP_CERT */
+        if (!strcmp(SSL_state_string(s), "TRCCC") && seen != NULL)
+            *seen = 1;
+    } else {
+        /* TLS_ST_CR_COMP_CERT */
+        if (!strcmp(SSL_state_string(s), "TRSCC") && seen != NULL)
+            *seen = 1;
+    }
+}
+
+static int test_ssl_cert_comp(int test)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    int expected_client = TLSEXT_comp_cert_none;
+    int expected_server = TLSEXT_comp_cert_none;
+    int client_seen = 0;
+    int server_seen = 0;
+    /* reverse default order */
+    int server_pref[] = { TLSEXT_comp_cert_zstd, TLSEXT_comp_cert_zlib, TLSEXT_comp_cert_brotli };
+    /* default order */
+    int client_pref[] = { TLSEXT_comp_cert_brotli, TLSEXT_comp_cert_zlib, TLSEXT_comp_cert_zstd };
+
```

Comment:
It might be nice to add some GREASE tests with undefined compression scheme IDs.